src/zjs_fs.c

```diff
@@ -0,0 +1,1048 @@
+
+    if (offset < 0 || length < 0 || position < 0) {
+        return invalid_args();
+    }
+    if (offset >= buffer->bufsize) {
+        return zjs_error("offset overflows buffer");
+    }
+    if (offset + length > buffer->bufsize) {
+        return zjs_error("offset + length overflows buffer");
+    }
+
+    // if a position was specified, seek to it before writing
+    if (fs_seek(&handle->fp, (uint32_t)position, SEEK_SET) != 0) {
+        return zjs_error("error seeking to position\n");
+    }
+
+    DBG_PRINT("writing to fp=%p, buffer=%p, offset=%lu, length=%lu\n", &handle->fp, buffer->buffer, (uint32_t)offset, (uint32_t)length);
+
+    uint32_t written = fs_write(&handle->fp, buffer->buffer + (uint32_t)offset, (uint32_t)length);
+
+    if (written != (uint32_t)length) {
+        DBG_PRINT("could not write %lu bytes, only %lu were written\n", (uint32_t)length, written);
+    }
+
+#ifdef ZJS_FS_ASYNC_APIS
+    if (async) {
+        jerry_value_t args[3];
+
+        args[0] = jerry_create_number(0);
+        args[1] = jerry_create_number(written);
+        args[2] = argv[1];
+
+        zjs_callback_id id = zjs_add_callback_once(js_cb,
+                                                   this,
+                                                   NULL,
+                                                   NULL);
+
+        zjs_signal_callback(id, args, sizeof(jerry_value_t) * 3);
+
+        return ZJS_UNDEFINED;
+    }
+#endif
+    return jerry_create_number(written);
+}
+
+static jerry_value_t zjs_write_sync(const jerry_value_t function_obj,
+                                    const jerry_value_t this,
+                                    const jerry_value_t argv[],
+                                    const jerry_length_t argc) {
+    return zjs_write(function_obj, this, argv, argc, 0);
+}
+
+#ifdef ZJS_FS_ASYNC_APIS
+static jerry_value_t zjs_write_async(const jerry_value_t function_obj,
+                                     const jerry_value_t this,
+                                     const jerry_value_t argv[],
+                                     const jerry_length_t argc) {
+    return zjs_write(function_obj, this, argv, argc, 1);
+}
+#endif
+
+static jerry_value_t zjs_truncate(const jerry_value_t function_obj,
+                                  const jerry_value_t this,
+                                  const jerry_value_t argv[],
+                                  const jerry_length_t argc,
+                                  uint8_t async)
+{
+    fs_file_t fp;
+    if (!jerry_value_is_number(argv[1])) {
+        return invalid_args();
+    }
+    if (async) {
+        if (!jerry_value_is_function(argv[2])) {
+            return invalid_args();
+        }
+    }
+    if (jerry_value_is_object(argv[0])) {
+        file_handle_t* handle;
+        if (!jerry_get_object_native_handle(argv[0], (uintptr_t*)&handle)) {
+            return zjs_error("native handle not found");
+        }
+        fp = handle->fp;
```

Comment:
For now, if supporting only 'r' and 'w' as I suggested, truncate should only be allowed if it was opened for 'w'. Later, it would be okay for 'r+' or 'w+' too. Not sure about a/a+, either not supported or only if truncate length is >= current seek position I guess.